draft-pauly-quic-datagram.md

```diff
@@ -179,40 +173,76 @@ Datagram Data:
-
-Although the DATAGRAM frame is not retransmitted upon loss detection,
-it does contribute to the maximum data for the overall connection. Packets that
-contain only DATAGRAM frames do need to be acknowledged, but
-implementations SHOULD defer and batch acknowledgements since
-the timing of these acknowledgements is not used for loss recovery.
-
-The DATAGRAM frame does not provide any explicit flow control signaling
-apart from the connection-level flow control. DATAGRAM frames are flow
-controlled only when the maximum data for the connection is hit, at which point
-the BLOCKED frame is sent.
-
-In cases in which a DATAGRAM frame is blocked due to connection-level
-flow control or congestion control, an implementation MAY drop the frame
-without sending it.
+## Flow Identifiers {#flow-id}
+
+Flow identifiers represent bidirectional flows of datagrams within a single QUIC
+connection. These are effectively equivalent to UDP ports, that allow basic
+demultiplexing of application data. Whenever one side of a connection
+receives a frame with a Flow ID was was not previously known, it MAY represent
+this to the application as a new flow of datagrams.
+
+The primary role of the QUIC transport towards the flow identifier is to provide
+a standard mechanism for demultiplexing application data flows, which may be
+destined for different processing threads in the application, akin to UDP sockets.
+
+Beyond this, a sender SHOULD ensure that DATAGRAM frames within a single flow
+are transmitted in order relative to one another. If multiple DATAGRAM frames can
+packed into a single packet, the sender SHOULD group them by Flow ID to
+promote fate-sharing within a specific flow and improve the ability to process batches
+of datagram messages efficiently on the receiver.
+
+Applications that do not have a need for the Flow ID can use the value zero on
+their DATAGRAM frames and clear the FLOW_ID bit to omit sending the identifier
+over the wire. If an application uses a mixture of DATAGRAM frames with and
+without the FLOW_ID bit set, the frames without it are assumed to be part of the
+application-level flow with Flow ID zero.
+
+## Acknowledgements
+
+Although DATAGRAM frames are not retransmitted upon loss detection, they are
+ack-eliciting ({{!I-D.ietf-quic-recovery}}). Receivers SHOULD support delaying
+ACK frames in reponse to receiving packets that only contain DATAGRAM frames,
+since the timing of these acknowledgements is not used for loss recovery.
+
+If a sender detects that a packet containing a specific DATAGRAM frame has been
+lost, the implementation MAY notify the application that the datagram was lost.
+Similarly, if a packet containing a DATAGRAM frame is acknowledged, the
+implementation MAY notify the application that the datagram was successfully
+transmitted and received.
+
+## Flow Control
+
+DATAGRAM frames do not provide any explicit flow control signaling,
+and do not contribute to any per-flow or connection-wide data limit.
```

Comment:
Yay!  I've been finding it hard to understand how connection wide flow control would work in the presence of loss...